extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerClient.java

```diff
@@ -72,7 +72,7 @@ ListenableFuture<Void> postResultPartitionBoundaries(
    * kind of unrecoverable exception).
    */
   ListenableFuture<Boolean> fetchChannelData(
-      String workerTaskId,
+      int workerNumber,
```

Comment:
IMO, the logic of which worker task to contact should be built outside this client as we would never want half the requests going to older workers, when worker fault tolerance is there, and half of them to the new workers. 
Hence, I think we should not touch `extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerClient.java` this class. 

Thoughts?

Reply:
I think `WorkerClient` should be refactored to include the worker number to save callers the complexity of resolving the ID to number. In the current indexer's implementation, this is an implementation detail that allows the logic to be wrapped up in the implementation class. 
The behavior based on fault tolerance would change in the controller task only (as the worker tasks won't be communicating with each other in fault-tolerant mode). In the controller task, irrespective of whether the fault tolerance is enabled, this would resolve to whatever is present in the `MSQWorkerTaskLauncher`, which would ultimately be responsible for determining where the call to a specific worker number should resolve.

Reply:
The worker-client needs to be a simple client. Given a taskId, do the required operation. 
I think that would be a more manageable contract for the users of the client as it is deterministic. 
When the msqWokerTaskLauncher changes something in the background, we do not want an operation to succeed on a worker in the wrong state. 

Users will always have to look up the taskID and get the worker state which is on the worker number. 
As we have async api's it can be very well possible that we end up getting success from a task that according to the controller state is in retrying. ie the worker is in retrying. 
In such a case we simply have to ignore the state change which becomes hard if we have only the worker id as worker states are on the worker number. 

Another reason which tipped me over towards the taskID approach is that worker-side client interface changes are not required in the worker JVM.